fedn/network/clients/client_api.py

```diff
@@ -0,0 +1,321 @@
+import enum
+import os
+import time
+from io import BytesIO
+from typing import Any, Tuple
+
+import requests
+
+import fedn.network.grpc.fedn_pb2 as fedn
+from fedn.common.config import FEDN_AUTH_SCHEME, FEDN_PACKAGE_EXTRACT_DIR
+from fedn.common.log_config import logger
+from fedn.network.clients.grpc_handler import GrpcHandler
+from fedn.network.clients.package_runtime import PackageRuntime
+from fedn.utils.dispatcher import Dispatcher
+
+
+class GrpcConnectionOptions:
+    def __init__(self, status: str, host: str, fqdn: str, package: str, ip: str, port: int, helper_type: str):
+        self.status = status
+        self.host = host
+        self.fqdn = fqdn
+        self.package = package
+        self.ip = ip
+        self.port = port
+        self.helper_type = helper_type
+
+
+# Enum for respresenting the result of connecting to the FEDn API
+class ConnectToApiResult(enum.Enum):
+    Assigned = 0
+    ComputePackgeMissing = 1
+    UnAuthorized = 2
+    UnMatchedConfig = 3
+    IncorrectUrl = 4
+    UnknownError = 5
+
+
+def get_compute_package_dir_path():
+    result = None
+
+    if FEDN_PACKAGE_EXTRACT_DIR:
+        result = os.path.join(os.getcwd(), FEDN_PACKAGE_EXTRACT_DIR)
+    else:
+        dirname = +"compute-package-" + time.strftime("%Y%m%d-%H%M%S")
+        result = os.path.join(os.getcwd(), dirname)
+
+    if not os.path.exists(result):
+        os.mkdir(result)
+
+    return result
+
+
+class ClientAPI:
+    def __init__(self):
+        self._subscribers = {"train": [], "validation": []}
+        path = get_compute_package_dir_path()
+        self._package_runtime = PackageRuntime(path)
+
+        self.dispatcher: Dispatcher = None
+        self.grpc_handler: GrpcHandler = None
+
+    def subscribe(self, event_type: str, callback: callable):
+        """Subscribe to a specific event."""
+        if event_type in self._subscribers:
+            self._subscribers[event_type].append(callback)
+        else:
+            raise ValueError(f"Unsupported event type: {event_type}")
+
+    def notify_subscribers(self, event_type: str, *args, **kwargs):
```

Comment:
I don't understand, what subscribers might a client have?

Reply:
@ahellander if you look in the client_v2.py file you can see how it is used. The overall idea is that client_api exposes functionality that a client can then make use of... In this particular example the client subscribes to events raised from the client_api. Events include training, validation or inference.

Reply:
I will add a little example too.